processor/src/main/java/org/jboss/logging/processor/apt/report/ReportWriter.java

```diff
@@ -129,41 +136,48 @@ protected String getLogLevel(final MessageMethod method) {
-     *
-     * @param messageInterface the message interface to get the methods for
-     *
-     * @return a sorted collection of message methods
-     */
-    protected Collection<MessageMethod> getSortedMessageMethods(final MessageInterface messageInterface) {
-        // Ensure the messages are sorted by the id
-        final List<MessageMethod> messageMethods = new ArrayList<>(messageInterface.methods());
-        Collections.sort(messageMethods, MessageIdComparator.INSTANCE);
-        return Collections.unmodifiableCollection(messageMethods);
-    }
+    String getUrl(final MessageMethod messageMethod, final String id) {
+        final ResolutionDoc resolutionDoc = getResolutionDoc(messageMethod);
+        if (resolutionDoc == null || resolutionDoc.skip() || DEFAULT_ID.equals(id)) {
+            return "";
+        }
 
-    /**
-     * Creates a {@linkplain java.util.Formatter format} string for creating message id's.
-     *
-     * @param messageInterface the message interface to create the format string for
-     *
-     * @return the format string
-     */
-    static String createMessageIdFormat(final MessageInterface messageInterface) {
-        final int idLen = messageInterface.getIdLength();
-        if (idLen > 0) {
-            return messageInterface.projectCode() + "%0" + messageInterface.getIdLength() + "d";
+        final StringBuilder result = new StringBuilder();
+
+        String base = baseUrl == null ? "" : baseUrl;
+        final String path = resolutionDoc.path();
+        final String suffix = resolutionDoc.suffix();
+        final String url = resolutionDoc.url();
+        if (!url.isEmpty()) {
+            base = url;
         }
-        return messageInterface.projectCode() + "%d";
-    }
 
-    static class MessageIdComparator implements Comparator<MessageMethod> {
-        static final MessageIdComparator INSTANCE = new MessageIdComparator();
+        if (!base.isEmpty()) {
+            result.append(base);
+            if (!base.endsWith("/") && !base.endsWith("#")) {
+                result.append('/');
+            }
+        }
+
+        if (path.isEmpty()) {
+            result.append(id);
+        } else {
+            result.append(path);
+        }
+
+        if (suffix.isEmpty()) {
+            result.append(getReportType().getExtension());
+        } else {
+            result.append(suffix);
+        }
+
+        return Expressions.resolve(messageInterface.expressionProperties(), result.toString());
+    }
 
-        @Override
-        public int compare(final MessageMethod o1, final MessageMethod o2) {
-            return Integer.compare(o1.message().id(), o2.message().id());
+    private ResolutionDoc getResolutionDoc(final MessageMethod messageMethod) {
+        if (messageMethod.isAnnotatedWith(ResolutionDoc.class)) {
+            return messageMethod.getAnnotation(ResolutionDoc.class);
         }
+        return messageInterface.getAnnotation(ResolutionDoc.class);
```

Comment:
It might make sense here to allow attributes defined on the type to be inherited. However it's not that difficult to just override these so maybe that doesn't make sense.